### PR TITLE
Inject full redirect URI into the logged out template to prevent reconciliation error

### LINKED ISSUE #13376: Themes: Fix Reconciliation Error
To repro:
* Run Calypso locally (error is hidden in production)
* In an incognito window, land at http://calypso.localhost:3000/themes
* Check the console:

image

This was probably caused by https://github.com/Automattic/wp-calypso/pull/13010

### PR DESCRIPTION
**Pull request by oigewan**
**Bug:**
As mentioned in #13376, the output of the server rendering of the logged out `/themes` page no longer matches the client rendered output.  This is due to the redirectUrl parameter that was added to the login URL in #13010.  Since the server doesn't have access to the `window` object, that link will always be different.

**Changes:** 
Since we don't have access to the `window` object, we'll have to manually construct the full URL of the page (the client uses `window.location.href`) and inject that into the logged out layout.  Unfortunately, the request's `originalUrl` property doesn't include the protocol, host, or port, so I had to construct this URL manually.

Here's how I chose to do this: 
1. Add a new property called `fullOriginalUri` on the request's `context` object.  This is done inside of `getDefaultContext`, but I'm happy to move this logic elsewhere if that's inappropriate.
2. Inject the value of that property into the layout component in the `makeLayoutMiddleware` as a prop named `redirectUri`.
3. Pass that all the way down to the `MasterbarLoggedOut` component and output it if it exists and the `window` does not.

**Notes:**
This my first PR for this project.  I tried to read all relevant documentation, but I still might have done something stupid.  If so, please let me know and point me to the relevant docs.

Fixes #13376 

cc: @ockham @klimeryk

### CODE REVIEW

**Comment by oigewan** [comment]
NOTE: I thought about just using `request.headers.host` to build up the redirect URL, but I decided to go with the method used to build up the redirect URL inside of the setUpLoggedInRoute method later `server/pages/index.js`.

File: server/pages/index.js [hunk 1]
```diff
@@ -128,6 +128,13 @@ function getDefaultContext( request ) {
 		initialServerState = getInitialServerState( serializeCachedServerState );
 	}
 
+	const protocol = request.get( 'X-Forwarded-Proto' ) === 'https' ? 'https' : 'http';
```
Comment (ockham) [inline server/pages/index.js]: Can't we use `request.protocol` here? (Disclaimer -- I just did a quick search that turned up [this SO answer](http://stackoverflow.com/a/10185427).)

**Comment by ockham** [comment]
This is awesome, thanks a lot for the fix @oigewan!

**Comment by ockham** [comment]
One thing that'd be nice to do a bit more elegantly is the passing of props down so many component levels (especially something as specific as `fullOriginalUri` -- when compared to pervasively used props such as our `primary` and `secondary` component trees).

By definition, we can only run into reconciliation errors in our isomorphic sections -- the ones that are server-rendered at all. For them, we have a helper that already adds [some attributes](https://github.com/Automattic/wp-calypso/blob/55f5504a51bcfa20f1ea2483aade64005e5b8eec/server/isomorphic-routing/index.js#L54) from Express' `request` to our "fake" `page.js`-style `context`. We could add whatever fields are still required and then build `fullOriginalUri` inside `client/controller` -- ideally leveraging the fact that we have a server-specific `index.node.js` and a client-specific `index.web.js` there so we don't even need the conditional in `masterbar/logged-out#getLoginUrl`, but are able to get the correct `redirectUri` for either case passed from `client/controller`.

Hope this makes sense -- don't hesitate to ask if it doesn't!

File: server/pages/index.js (hunk 1 above)
Comment (oigewan) [inline server/pages/index.js]: Like I mentioned in my earlier comment, I thought about using properties of the request to build up the URL.  However, I saw how the login redirect URL was being built up in the setUpLoggedInRoute.  Specifically, right here: https://github.com/Automattic/wp-calypso/blob/87f4b5c35f6d7f56a01c292a7bdc2534d4b72a26/server/pages/index.js#L211

I thought that it'd be better to use the same kinds of method to build up the redirect URI here.

Comment (oigewan) [inline server/pages/index.js]: I can use that method to get the protocol if you want, I just wanted to point out that we already build up a very similar URL elsewhere in the same file.  Would seem odd to use two different methods of creating the URL.  If we're going to use `request.protocol` here, I'd suggest using it in the other location as well.

**Comment by oigewan** [comment]
@ockham I can see what you're saying about building the URL inside of the two different controllers and always passing down the URL that we want, and it does make sense.  However, I would point out that `client/controller` doesn't actually have access to the context.  Only the `makeLayoutMiddleware` method inside of `client/controller/shared` has access to that.

If you'd rather go this route, I can add request properties onto the context inside of `getEnhancedContext`, move the `if ( typeof window === 'undefined' ) {` check out of the logged out masterbar component and into the `makeLayoutMiddleware` function, then unconditionally pass in that `redirectUri` prop either as `window.location.href` or the built URL based on the enhanced context.  How does that sound?

**Comment by oigewan** [comment]
@ockham PR updated as previously mentioned.  I rebased this commit, but I stashed the previous version of this PR so that it can be restored if we prefer that.

**Comment by ockham** [comment]
Yes, perfect!

I noticed there's a regression for non-isomorphic sections, e.g. for logged-out `/start`, where the redirect URI is no longer appended to the login link, which is due to `MasterbarLoggedOut` being rendered from `client/layout` (as opposed to `layout/logged-out`). (`client/layout` OTOH is rendered from `client/boot/wordpress-com` for non-isomorphic sections.)

As much as I hate it, I think this means we'll have to have `redirectUri` fall back to `window.location.href` deeper down in the component hierarchy (possibly inside `MasterbarLoggedOut` after all), where both possible execution branches are ensured to work.

**Comment by oigewan** [comment]
@ockham I was kinda afraid of that when I started coding.  Oh well.  Rebased and pushed.

**Comment by oigewan** [comment]
Also thanks for putting up with the delay in responses.  I live in New Zealand and I have a day job, so...

**Comment by ockham** [comment]
> Also thanks for putting up with the delay in responses. I live in New Zealand and I have a day job, so...

Thank _you_ for putting up with my review!

This is working great now, so I'm going to merge it.

**Comment by klimeryk** [comment]
Thanks for the fix, @oigewan! 🙇

**Pull request merged**